src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionXmlFormatReader.cs

```diff
@@ -88,33 +115,75 @@ private void ReflectionReadMembers(object obj, XmlReaderDelegator xmlReader, Xml
+        // Put all members of the contract (including types from base contract) into 'members'.
+        private int ReflectionGetMembers(ClassDataContract classContract, DataMember[] members)
         {
-            int memberCount = 0;
+            int memberCount = (classContract.BaseContract == null) ? 0 : ReflectionGetMembers(classContract.BaseContract, members);
+            int childElementIndex = memberCount;
+            for (int i = 0; i < classContract.Members.Count; i++, memberCount++)
+            {
+                members[childElementIndex + i] = classContract.Members[i];
+            }
+
+            return memberCount;
+        }
+
+        private void ReflectionReadMember(object obj, int memberIndex, XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, DataMember[] members)
+        {
+            DataMember dataMember = members[memberIndex];
 
-            Type classType = _classContract.UnadaptedClassType;
-            int i = memberIndex + memberCount;
-            DataMember dataMember = _classContract.Members[i];
             Type memberType = dataMember.MemberType;
+            if (dataMember.IsGetOnlyCollection)
+            {
+                var memberValue = ReflectionGetMemberValue(obj, dataMember.MemberInfo);
+                context.StoreCollectionMemberInfo(memberValue);
+                ReflectionReadValue(memberType, dataMember.Name, _classContract.StableName.Namespace);
+            }
+            else
+            {
+                var value = ReflectionReadValue(memberType, dataMember.Name, _classContract.StableName.Namespace);
+                MemberInfo memberInfo = dataMember.MemberInfo;
+                if (memberInfo != null)
+                {
+                    ReflectionSetMemberValue(obj, value, memberInfo);
+                }
+                else
+                {
+                    throw new NotImplementedException("PropertyInfo incorrect");
+                }
+            }
+        }
 
-            var value = ReflectionReadValue(memberType, dataMember.Name, _classContract.StableName.Namespace);
-            MemberInfo memberInfo = dataMember.MemberInfo;
-            if (memberInfo != null)
+        private object ReflectionGetMemberValue(object obj, MemberInfo memberInfo)
+        {
+            object memberValue = null;
+            if (memberInfo is PropertyInfo)
+            {
+                PropertyInfo propInfo = (PropertyInfo)memberInfo;
+                memberValue = propInfo.GetValue(obj);
+            }
+            else if (memberInfo is FieldInfo)
             {
-                ReflectionSetMemberValue(obj, value, memberInfo);
+                FieldInfo fieldInfo = (FieldInfo)memberInfo;
+                memberValue = fieldInfo.GetValue(obj);
             }
             else
             {
-                throw new NotImplementedException("PropertyInfo incorrect");
+                throw new NotImplementedException("Unknown member type");
```

Comment:
Should this be InvalidOperationException?  I'm a little confused by NotImplementedException is being used to catch programming errors.